leshan-core/src/main/java/org/eclipse/leshan/core/request/execute/Arguments.java

```diff
@@ -0,0 +1,260 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Orange.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Orange - Add better support for Arguments of Execute Operation.
+ *******************************************************************************/
+package org.eclipse.leshan.core.request.execute;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.leshan.core.util.StringUtils;
+
+/**
+ * Arguments for Execute Operation.
+ */
+public class Arguments implements Iterable<Argument> {
+
+    private final List<Argument> argumentList;
+
+    /**
+     * Arguments should be build using {@link ArgumentsBuilder}.
+     */
+    private Arguments(List<Argument> argumentList) {
+        this.argumentList = argumentList;
```

Comment:
I don't see any information about Argument should be unique. That is a reason why I used List instead of Map and it's a reason why I don't provided access argument by digit. I established that arguments are not unique and client implementation decide how interpret two arguments with same digit.

Or I'm wrong?

Reply:
Good point.
My initial understanding was pretty clear that each arguments are defined by 1 digit key.
I double checked and you're right there is nothing clear about that. :grimacing: 

There is just maybe a kind of hint at [Table: D.1-3 Executable Resource Arguments Definition](http://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#Table-D1-3-Executable-Resource-Arguments-Definition)

ID | Resource Name | Key | Name | Type | Range or Enum | Unit | Description
-- | -- | -- | -- | -- | -- | -- | --
  |   | [0:9] | String | Data Types | If any | If any |  

As column are named **Key**. :thinking: 

Here the example in the spec about [portfolio](http://www.openmobilealliance.org/release/LWM2M_PortfolioObj/V1_0_1-20200616-A/OMA-TS-LWM2M_PortfolioObj-V1_0_1-20200616-A.pdf) but again this is only hints... :confused.

To be honest I'm not sure to see what would be the benefits to propose an arguments API like : `0='arg1',0='arg2',0='arg3'` instead of `0='arg1,arg2,arg3'`. But again this is only speculation.

So, I just asked at OMA https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/536 but it could be long to get answer (and some time we don't even get it) so we should find a short-term solution. 

So either : 
-  we keep the more open solution with maybe a no so good way to get a specific arguments.
- or we move to map and we bet that this is not allowed to have same digit twice.
- or a kind of mixed we keep the list and we provide a way to get argument by digit anyway.

Reply:
> I'm afraid adding getByDigit() (or something like that) method in case of more items at same digit. Or maybe I just overthinking?
>
> And what about getFirstByDigit()? We can implement desired functionality and if we establish that we allow only one digit per value (it would be unique), then we can rename method to e.g. getByDigit() (or just get())

Yep, I understand. 

Related to this issue there is also the `toMap` methods which should behave strangely if a digit is used more than once. 

My bet is that "reusing digit more than once" is either unexpected or will be rarely used. So I will go with either 2) or 3).
If we go with 3),  I propose to provide a straightforward API (pretty much the same than if we have chosen  2)) and add some javadoc to explain the behavior in case same digit is used twice (e.g. only first one is used).

The drawback with 3) the API will allow to create Arguments "reusing digit more than once", if we think we want to tolerate it but not encourage it maybe the Builder should not allow this (at all or by default) ?

Reply:
The solution 2) looks like the simplest. I'll add uniqueness validation and method access by digit.

Reply:
Let's go for 2) :slightly_smiling_face: